### PR TITLE
Support yolox pose refer from mmpose

### PR DESCRIPTION
**Pull request by yechenzhi**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation for this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a potential influence on downstream projects, this PR should be tested with downstream projects, like MMDetection or MMClassification.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: configs/_base_/datasets/coco.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
+dataset_info = dict(
```
Comment (hhaAndroid) [inline configs/_base_/datasets/coco.py:1]: configs/_base_/datasets/coco.py -> configs/_base_/pose//coco.py

File: configs/yolox/pose/yolox-pose_l_4xb64-300e_coco.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+_base_ = ['./yolox-pose_s_8xb32-300e_coco.py']
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_l_4xb64-300e_coco.py:1]: 先不加l的

File: configs/yolox/pose/yolox-pose_m_4xb64-300e_coco.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+_base_ = ['./yolox-pose_s_8xb32-300e_coco.py']
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_m_4xb64-300e_coco.py:1]: configs/yolox/pose/yolox-pose_m_4xb64-300e_coco.py -> configs/yolox/pose/yolox-pose_m_8xb32-300e-rtmdet-hyp_coco.py

File: configs/yolox/pose/yolox-pose_s_8xb32-300e_coco.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
+_base_ = ['../yolox_s_fast_8xb8-300e_coco.py']
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_s_8xb32-300e_coco.py:1]: configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py

Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_s_8xb32-300e_coco.py:1]:
```suggestion
_base_ = ['../yolox_s_fast_8xb32-300e-rtmdet-hyp_coco.py']
```

File: mmyolo/datasets/pose_coco.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+from typing import Any
+
+from mmengine.dataset import force_full_init
+from mmpose.datasets import CocoDataset as MMPoseCocoDataset
+
+from ..registry import DATASETS
+
+
+@DATASETS.register_module()
+class CocoPoseDataset(MMPoseCocoDataset):
```
Comment (hhaAndroid) [inline mmyolo/datasets/pose_coco.py:11]:
```suggestion
class PoseCocoDataset(MMPoseCocoDataset):
```

File: mmyolo/datasets/transforms/bbox_keypoint_structure.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+from mmdet.structures.bbox import HorizontalBoxes
+from torch import Tensor
+
+DeviceType = Union[str, torch.device]
+T = TypeVar('T')
+IndexType = Union[slice, int, list, torch.LongTensor, torch.cuda.LongTensor,
+                  torch.BoolTensor, torch.cuda.BoolTensor, np.ndarray]
+
+
+class BBoxKeypoints(HorizontalBoxes):
```
Comment (hhaAndroid) [inline mmyolo/datasets/transforms/bbox_keypoint_structure.py:16]: 这个封装对象是为了mmpose 简化用的，mmyolo 不需要。只需要更改所有pipeline，支持 keypoint 就行

File: mmyolo/datasets/transforms/transforms.py [hunk 7]
```diff
@@ -1555,3 +1558,119 @@ def transform(self, results: dict) -> dict:
         results['gt_bboxes'] = self.box_type(
             results['gt_bboxes'].regularize_boxes(self.angle_version))
         return results
+
+
+@TRANSFORMS.register_module()
+class PoseToDetConverter(BaseTransform):
```
Comment (hhaAndroid) [inline mmyolo/datasets/transforms/transforms.py:1564]: 这个咱们不需要

File: mmyolo/datasets/transforms/transforms.py [hunk 8]
```diff
@@ -1555,3 +1558,119 @@ def transform(self, results: dict) -> dict:
...
+        )
+
+        results['gt_ignore_flags'] = np.array([False] * num_instances)
+        results['gt_bboxes_labels'] = np.array(results['category_id']) - 1
+
+        return results
+
+
+@TRANSFORMS.register_module()
+class PackDetPoseInputs(PackDetInputs):
```
Comment (hhaAndroid) [inline mmyolo/datasets/transforms/transforms.py:1598]: 直接修改 mmyolo 里面的 PackDetInputs 就行吧

File: mmyolo/datasets/transforms/transforms.py [hunk 9]
```diff
@@ -1555,3 +1558,119 @@ def transform(self, results: dict) -> dict:
...
+        # which is necessary for `PackDetInputs` but not guaranteed during
+        # inference with an inferencer
+        for key in self.meta_keys:
+            if key not in results:
+                results[key] = None
+        return super().transform(results)
+
+
+@TRANSFORMS.register_module()
+class FilterDetPoseAnnotations(FilterDetAnnotations):
```
Comment (hhaAndroid) [inline mmyolo/datasets/transforms/transforms.py:1630]:
```suggestion
class FilterAnnotations(FilterDetAnnotations):
```

File: mmyolo/models/data_preprocessors/data_preprocessor.py [hunk 10]
```diff
@@ -300,3 +301,29 @@ def _get_random_size_and_interp(self) -> Tuple[int, int]:
         else:
             interp_mode = None
         return input_size, interp_mode
+
+
+@MODELS.register_module()
+class PoseBatchSyncRandomResize(BatchSyncRandomResize):
```
Comment (hhaAndroid) [inline mmyolo/models/data_preprocessors/data_preprocessor.py:307]: 直接修改mmyolo 这个类就行

File: mmyolo/datasets/transforms/transforms.py (hunk 8 above)
Comment (yechenzhi) [reply mmyolo/datasets/transforms/transforms.py:1598]: PackDetInputs 是mmdet里的，所以还是需要？

File: mmyolo/models/data_preprocessors/data_preprocessor.py (hunk 10 above)
Comment (yechenzhi) [reply mmyolo/models/data_preprocessors/data_preprocessor.py:307]: BatchSyncRandomResize 也是mmdet的

Comment (yechenzhi) [reply mmyolo/models/data_preprocessors/data_preprocessor.py:307]: 哦我看到了YOLOXBatchSyncRandomResize，ignore me,,

File: configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+model = dict(
+    type='YOLODetector',
+    init_cfg=dict(
+        _delete_=True,
+        type='Pretrained',
+        checkpoint='https://download.openmmlab.com/mmyolo/v0/yolox/'
+        'yolox_s_fast_8xb32-300e-rtmdet-hyp_coco/yolox_s_fast_'
+        '8xb32-300e-rtmdet-hyp_coco_20230210_134645-3a8dfbd7.pth'),
+    data_preprocessor=dict(
+        type='mmdet.DetDataPreprocessor',
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py:13]: YOLOv5DetDataPreprocessor

File: configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+        pad_size_divisor=32,
+        batch_augments=[
+            dict(
+                type='PoseBatchSyncRandomResize',
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py:17]: 直接修改现有代码就行，不需要新增类

File: configs/yolox/pose/yolox-pose_m_8xb32-300e-rtmdet-hyp_coco.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+        deepen_factor=0.67,
+        widen_factor=0.75,
+    ),
+    neck=dict(
+        deepen_factor=0.67,
+        widen_factor=0.75,
+    ),
+    bbox_head=dict(head_module=dict(widen_factor=0.75)))
+
+train_dataloader = dict(batch_size=64)
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_m_8xb32-300e-rtmdet-hyp_coco.py:18]: 应该是32吧

File: configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+    dict(type='FilterAnnotations', by_keypoints=True, keep_empty=False),
+    dict(
+        type='PackDetPoseInputs',
+        meta_keys=('img_id', 'img_path', 'ori_shape', 'img_shape'))
+]
+
+train_pipeline_stage2 = [
+    *pre_transform,
+    dict(type='Resize', scale=img_scale, keep_ratio=True),
+    dict(
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py:84]: 用这个不行把，没有处理keypoint

File: configs/yolox/pose/yolox-pose_tiny_4xb64-300e_coco.py [hunk 15]
```diff
@@ -0,0 +1,68 @@
+_base_ = ['./yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py']
+
+# model settings
+model = dict(
+    init_cfg=dict(checkpoint='https://download.openmmlab.com/mmyolo/v0/yolox/'
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_tiny_4xb64-300e_coco.py:5]: Load from

File: configs/yolox/pose/yolox-pose_tiny_4xb64-300e_coco.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+                  'yolox_tiny_fast_8xb32-300e-rtmdet-hyp_coco/yolox_tiny_fast_'
+                  '8xb32-300e-rtmdet-hyp_coco_20230210_143637-4c338102.pth'),
+    data_preprocessor=dict(batch_augments=[
+        dict(
+            type='PoseBatchSyncRandomResize',
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_tiny_4xb64-300e_coco.py:10]: 同样问题

File: mmyolo/models/task_modules/assigners/pose_sim_otaa_assigner.py [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+from torch import Tensor
+
+from mmyolo.registry import MODELS, TASK_UTILS
+
+INF = 100000.0
+EPS = 1.0e-7
+
+
+@TASK_UTILS.register_module()
+class PoseSimOTAAssigner(SimOTAAssigner):
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/pose_sim_otaa_assigner.py:18]: 可以直接继承修改，而不用区分pose还是det吗

File: mmyolo/models/utils/misc.py [hunk 18]
```diff
@@ -95,3 +97,90 @@ def gt_instances_preprocess(batch_gt_instances: Union[Tensor, Sequence],
                                          device=batch_gt_instances.device)
 
         return batch_instance
+
+
+class OutputSaveObjectWrapper:
```
Comment (hhaAndroid) [inline mmyolo/models/utils/misc.py:102]: 用于干啥的

File: configs/yolox/pose/yolox-pose_m_8xb64-300e-rtmdet-hyp_coco.py [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+        deepen_factor=0.67,
+        widen_factor=0.75,
+    ),
+    neck=dict(
+        deepen_factor=0.67,
+        widen_factor=0.75,
+    ),
+    bbox_head=dict(head_module=dict(widen_factor=0.75)))
+
+train_dataloader = dict(batch_size=64)
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_m_8xb64-300e-rtmdet-hyp_coco.py:17]: m模型为啥是 64，能否统一全部用 bs32

File: configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py [hunk 20]
```diff
@@ -0,0 +1,188 @@
+_base_ = ['../yolox_s_fast_8xb32-300e-rtmdet-hyp_coco.py']
+
+load_from = 'https://download.openmmlab.com/mmyolo/v0/yolox/yolox_s_fast_8xb32-300e-rtmdet-hyp_coco/yolox_s_fast_8xb32-300e-rtmdet-hyp_coco_20230210_134645-3a8dfbd7.pth'  # noqa
+
+# model settings
+model = dict(
+    type='YOLODetector',
+    data_preprocessor=dict(
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py:8]: 这个和base完全一样，不用重写了

File: configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py [hunk 21]
```diff
@@ -0,0 +1,188 @@
+_base_ = ['../yolox_s_fast_8xb32-300e-rtmdet-hyp_coco.py']
+
+load_from = 'https://download.openmmlab.com/mmyolo/v0/yolox/yolox_s_fast_8xb32-300e-rtmdet-hyp_coco/yolox_s_fast_8xb32-300e-rtmdet-hyp_coco_20230210_134645-3a8dfbd7.pth'  # noqa
+
+# model settings
+model = dict(
+    type='YOLODetector',
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py:7]: 删掉

File: configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+                random_size_range=(480, 800),
+                size_divisor=32,
+                interval=1)
+        ]),
+    bbox_head=dict(
+        type='YOLOXPoseHead',
+        head_module=dict(
+            type='YOLOXPoseHeadModule',
+            num_classes=1,
+            num_keypoints=17,
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py:23]: 这个参数提到前面去

File: configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+            metainfo='configs/_base_/pose/coco.py',
+            loss_weight=30.0)),
+    train_cfg=dict(
+        assigner=dict(
+            type='PoseSimOTAAssigner',
+            center_radius=2.5,
+            iou_calculator=dict(type='mmdet.BboxOverlaps2D'),
+            oks_calculator=dict(
+                type='OksLoss', metainfo='configs/_base_/pose/coco.py'))),
+    test_cfg=dict(
```
Comment (hhaAndroid) [inline configs/yolox/pose/yolox-pose_s_8xb32-300e-rtmdet-hyp_coco.py:36]: 这个如果也没有改，就不用重写了

File: mmyolo/models/data_preprocessors/data_preprocessor.py [hunk 24]
```diff
@@ -16,6 +17,34 @@
                  None]
 
 
+@MODELS.register_module()
+class PoseBatchSyncRandomResize(BatchSyncRandomResize):
```
Comment (hhaAndroid) [inline mmyolo/models/data_preprocessors/data_preprocessor.py:21]: 不需要

File: mmyolo/models/dense_heads/yolox_pose_head.py [hunk 25]
```diff
@@ -0,0 +1,410 @@
...
+                        with_nms: bool = True) -> List[InstanceData]:
+        """Transform a batch of output features extracted by the head into bbox
+        and keypoint results.
+
+        In addition to the base class method, keypoint predictions are also
+        calculated in this method.
+        """
+
+        # calculate predicted bboxes and get the kept instances indices
+        with OutputSaveFunctionWrapper(
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/yolox_pose_head.py:270]: 解释下这两个装饰器作用

File: mmyolo/models/dense_heads/yolox_pose_head.py [hunk 26]
```diff
@@ -0,0 +1,410 @@
...
+            kpts = kpts[keep_idxs_nms]
+            kpts_vis = kpts_vis[keep_idxs_nms]
+            keep_indices_nms_idx += 1
+
+            pred_instances.keypoints = kpts
+            pred_instances.keypoint_scores = kpts_vis
+
+        return results_list
+
+    def predict(self,
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/yolox_pose_head.py:342]: 这个重写是否有必要？

File: mmyolo/models/task_modules/assigners/pose_sim_otaa_assigner.py [hunk 27]
```diff
@@ -0,0 +1,209 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/pose_sim_otaa_assigner.py:1]: mmyolo/models/task_modules/assigners/pose_sim_otaa_assigner.py -> mmyolo/models/task_modules/assigners/pose_sim_ota_assigner.py

File: mmyolo/datasets/transforms/transforms.py [hunk 28]
```diff
@@ -7,16 +7,21 @@
 import mmcv
 import numpy as np
 import torch
+from mmcv.image.geometric import _scale_size
```
Comment (hhaAndroid) [inline mmyolo/datasets/transforms/transforms.py:10]: 一些重点的类或者函数，需要写单元测试

File: mmyolo/models/dense_heads/yolox_pose_head.py (hunk 26 above)
Comment (yechenzhi) [reply mmyolo/models/dense_heads/yolox_pose_head.py:342]: mmpose里面计算指标时候是用的numpy的array，如果是tensor的话会报错

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**